### PR TITLE
New virtual node and renderer architecture

### LINKED ISSUE #5: Suggestion: fragment nodes
Have as an option to `Html` a `Fragment` constructor with same function as React `</>`.

### LINKED ISSUE #33: Snabbdom dependency
Currently, the library outsources dom patching to snabbdom. This has a few issues (that may or may not be worth being worked on):

* As an external npm dependency, it adds friction to the development process

* We can't update snabbdom at all since newer versions use ES modules which the PureScript compiler doesn't support yet

* These might be eased by the latest changes (e.g., keyed), somehow managing to upgrade snabbdom, or just some work around code in our part, but there are _many_ things that the current snabbdom implementation doesn't play nice with. From the top of my head: externally modifying nodes (in cases where hooks don't apply, usually with server side rendering), conditional rendering (e.g., if and else in views), content editable usage...

* There is something to be said about improving bundle size and performance, but I am still looking into these (and snabbdom doesn't really maintain benchmarks)

A couple solutions come to mind:

1. Simply wait out on the PS compiler, and then upgrade snabbdom

This is the easiest course of action, for obvious reasons. There is no time frame guarantee, though, or that all points will improve.

2. Fork snabbdom and iron it out for usage with the library

This one could solve most of the outlined problems. I don't know how much the associated maintenance cost would increase, however. 

3. Use something else

Likely a lot of work. We can just fork any other JavaScript library --  as long the library current user facing API stays the same -- no need to go full NIHS. Feature matching might be a concern (e.g., if the new implementation doesn't use a virtual dom), specially given all of @chekoopa recent work :upside_down_face:.

### LINKED ISSUE #39: Runtime error when input element uses the `list` attribute
I was attempting to build a `view` function that included an `<input>`. That `<input>` made use of a `<datalist>` so it would autocomplete. Here's a cut down example of what I was trying to do:

```PureScript
view :: Model -> Html Message
view model =
  HE.main "main"
    $ [ HE.label [ HA.for "broken-input" ] "Broken input"
      , HE.input [ HA.type' "search", HA.name "broken-input", HA.id "broken-input", HA.list "search-results", HA.value model.searchResult, HA.onInput SearchStuff ]
      , HE.datalist [ HA.id "search-results" ] [ HE.option_ "Search..." ]
      ]
```
On Chrome this generates a runtime error: `Uncaught TypeError: Cannot set property list of #<HTMLInputElement> which has only a getter`
On Firefox this generates the same error (with a different message): `Uncaught TypeError: setting getter-only property "list"`

Seems the problem is Flame creates the Input and attempts to set the `list` attribute after, but since it's read-only this cannot work.

### PR DESCRIPTION
**Pull request by easafe**
TL;DR: Replaces snabbdom (and snabbdom-html) with a new renderer for both server side and client side

Changes:

1. Virtual nodes

* There is no longer an intermediate representation for virtual nodes. The `Html` and `NodeData` types have been changed to foreign imports (well, `Html` could still be typed but I don't see the point). This significantly improves memory usage at the cost of losing referential transparency for virtual node creation (mainly the `node` property). As the user can only create virtual nodes with `Flame.Html.Element` functions this is safe 

* Classes are also merged like styles

* Fragment nodes have been added

* Lazy nodes (aka thunks) no longer required a tag; the key parameter is now `Maybe String`

* Key node data for keyed rendering

* New managed nodes to replace hooks

2. Client side rendering

* The new DOM renderer no longer destroys the root parent node (tho this can be changed if there are good reasons)

* Events are now synthetic and managed by the renderer

* Same event type can have multiple handlers

* The new keyed implementation performs much better than snabbdom 

* Using `innerHtml` is now completely safe (well, the diffing, one could still get XSS :)

* Conditional views are supported

* No more external dependencies

3. Server side rendering

* Faster hydration that only sets the `node` property (and events) on the `Html` tree

* No more external dependencies; bundle size is also somewhat smaller

4. Clean up

* Internal files (that shouldn't be depended on by users) have been moved to respective `Internal` folders

* `Flame.HTML` has been renamed to `Flame.Html` for consistency

* `Flame.External` is now `Flame.Html.Signal`

* `innerHTML` renamed to `innerHtml`

To consider:

1. There is a lot of new FFI code, some of which can't be made DRY due to PureScript not supporting modules

2. Maybe could be the time to define which browsers are supported or not

3. Might be worth to add `Eq` and `Show` instances back

4. I am not quite sure about the current lazy nodes implementation -- it uses a convenient `rendered` property to cache the node, but it feels hacky

5. There is a few things that should be addressed like safe usage of content editable, typing attributes a bit better, etc

Closes #33 and closes #5 and closes #34 and closes #39

### CODE REVIEW

**Comment by easafe** [comment]
@chekoopa

**Comment by chekoopa** [comment]
Trying for a quick review, I'll try building my code on this version later. 
A friendly note on reformatting: it should have been a separate commit, it's kinda difficult to scan diffs

The main point is it shouldn't have any regressions on rendering. Snabbdom elimination is neat, tho. Somewhere is looks closer to that enum-based code from Hedwig, but as it's depend-less...

Treating SVG nodes as `Html` is good (putting unnecessary details deeper into implementation) and faithful to previous interface.

Signal implementation (Flame.Html.Signal.Document etc.) seems dirty imo, as `channelHandler` seemingly could be overloaded inside of Purs code.

Hooks are just a way to implement a foreign-controlled node, which doesn't comply MVU architecture. And that particularly relies on changing `innerHtml` attribute. 
Either we can even implement a separate node type, [like in Elm](https://github.com/evancz/elm-markdown/blob/e395295cbe0b460ab051c4040011470628f05b72/src/Native/Markdown.js#L6-L43), or have an old-fashioned hook design (to be put into `createNode` (init/create), `updateNodeData` (update, pre/postpatch)). 

However, current implementation has no way to detect and process node destruction completely, only particular parent nodes, [which get `clearNode` called on](https://github.com/easafe/purescript-flame/blob/4971b5311938b787a01bb9d3e71b30060e314075/src/Flame/Renderer/Internal/Dom.js#L277). Can't put a finger on a particular use case of `destroy` hook, however, it might be useful for cleanup and stateful elements. Though, it probably will have an impact on performance (recursively search for destroy hooks and call them all?)

**Comment by easafe** [comment]
> A friendly note on reformatting: it should have been a separate commit, it's kinda difficult to scan diffs on that.

Oh I realize now I changed the identation :S 

> Signal implementation (Flame.Html.Signal.Document etc.) seems dirty imo, as channelHandler seemingly could be overloaded inside of Purs code

I didn't touch that one, but I guess we can supply the FFI with the signal functions

> Either we can even implement a separate node type, like in Elm

So this be a new node type `Html`with functions to render and update the dom node? This feels more fitting than snabbdom like hooks

> Can't put a finger on a particular use case of destroy hook

I think this is intended for modules/libraries. Don't see the need for it either

**Comment by easafe** [comment]
To replace hooks, I added `createHtml` with (simplified) type 
```purescript
{ 
  createElement :: arg -> Effect Element, 
  updateElement :: Element -> arg -> arg -> Effect Element 
} 
-> NodeData message 
-> arg 
-> Html message
```
with the following considerations:

* Server side rendering skips this kind of nodes, as element creation server side demands polyfills. For me that is reason enough to just use `innerHtml`

* I am not sure about using `Element` instead of `Node`. On one hand it doesn't make sense to return things like fragments, on the other it "demands" wrapping things (like text nodes) in a node

* As the node can carry data with the `arg` parameter seems fine to type it as `Effect` instead of expecting FFI code

I also moved keys and lazy nodes to the Html module to avoid repeating long chunks of FFI code.

**Comment by chekoopa** [comment]
Right, though it seems to me that `updateElement` should be able to change `Element` in place. Natively, they are mutable. Only proper mutable data implementation I know is `purescript-refs`.

**Comment by easafe** [comment]
I don't think you need refs. The first parameter of `updateElement` is the created node, which you can use the usual DOM functions.

**Comment by chekoopa** [comment]
Like, we recall to that idea of Purs-native element mapping, but now for real with `Web.DOM`? :thinking: Guess I'm in.

However, it'll take some sweetening and foreign functions (`Web.DOM` lacks functions for `innerHTML`).

**Comment by easafe** [comment]
Well, there is nothing preventing it to be FFI wholesale -- you could use `Effect.Uncurried.mkEffectFn3` and others. But I think it should be `Node` instead of `Element` after all.

I will push these changes later, together with code to support using `if else` in views.

**Comment by chekoopa** [comment]
As we can build wrappers over `ElementRenderer` (or, should I say, `NodeRenderer`) with `fromNode`, it's pretty possible. However, some sugar could be useful. Like, `ElementRenderer`-plus' `updateElement` unwraps `Node` to `Element` and if it's `Nothing`, it re-calls `createElement`.

**Comment by easafe** [comment]
> As we can build wrappers over ElementRenderer (or, should I say, NodeRenderer) with fromNode, it's pretty possible. However, some sugar could be useful. Like, ElementRenderer-plus' updateElement unwraps Node to Element and if it's Nothing, it re-calls createElement.

Not sure if I follow. Any kind of un/wrapping on the `*Element` functions will probably not be enough, since you `Element` is already an interface for html/svg elements.

Inner html and text on `Web.Dom` is quite handy, not sure if they will merge it tho.

**Comment by chekoopa** [comment]
> Any kind of un/wrapping on the `*Element` functions will probably not be enough, since `Element` itself is an interface for html/svg elements.

True. And these unwrappers only check whether a desired interface exists and a desired value implements it. Wrappers only do `unsafeCoerce`. So, `Node` is basically a parent class, and it can be processed either through safe wrappers or unsafe foreigns.
 
> Inner html and text on `Web.Dom` is quite handy, not sure if they will merge it tho.

They've thrown it away with 'not-in-a-spec'. I'm writing a foreign `Options -> String -> Node -> Effect Node`, then

**Comment by easafe** [comment]
Some more changes:

* Lots of fixes for hydration and conditional views (kind of a pain, I can see now why `snabbom` doesn't support it)
*  `createHtml` to `managed` as it makes more sense
*  Fail-safe for reused html objects: if the node property is already set, shallow copy that object. This makes the code a bit cumbersome in some parts, but it is a lot more efficient than blinding copying over the entire html tree for every update.

**Comment by chekoopa** [comment]
That's a lot of damage. :eyes: 

Well, Travis also needs some setting to bail on compilation failure.

```
Error found:
in module Examples.NoEffects.Counters.Main
at examples/NoEffects/Counters/Counters.purs:38:25 - 38:65 (line 38, column 25 - line 38, column 65)
  No type class instance was found for
                             
    Data.Functor.Functor Html
                             
while applying a function map
  of type Functor t0 => (t1 -> t2) -> t0 t1 -> t0 t2
  to argument CounterMessage index
while inferring the type of map (CounterMessage index)
in value declaration view
where t0 is an unknown type
      t2 is an unknown type
      t1 is an unknown type
```

**Comment by easafe** [comment]
Turns out something does use the `Functor` instance :s I will think of something, still looking through multiple handlers for a single event.

**Comment by chekoopa** [comment]
Not just something, but Html itself needs to map messages between types. That's a core feature, actually. 

Besides of Functor, Html also implemented Eq and Show, but I think Functor is essential.

https://github.com/easafe/purescript-flame/blob/flame_oh_hot_man/examples/NoEffects/Counters/Counters.purs#L38

**Comment by kim366** [comment]
I didn't read through everything, but personally I like this library BECAUSE I have access to Snabbdom Hooks. It may be true that it doesn't fit the MVU architecture, but the reason I went Purescript instead of Elm was more freedom. I am manipulating selections inside input fields and that is only possible in a robust manner with the postpatch hook.

**Comment by easafe** [comment]
Hello @kim366 , do you have any example code of something possible with postpatch hooks but not with the new API? https://github.com/easafe/purescript-flame/blob/flame_oh_hot_man/src/Flame/Html/Element.purs#L94

**Comment by kim366** [comment]
I just took a short look at the new API. I could try it in the next week. Am I understanding correctly that you are writing a new diffing algorithm from scratch?

**Comment by easafe** [comment]
The diffing algorithm is the same as other js libraries

keyed nodes (lifted from `stage0`): https://github.com/easafe/purescript-flame/blob/flame_oh_hot_man/src/Flame/Renderer/Internal/Dom.js#L409

non keyed nodes: https://github.com/easafe/purescript-flame/blob/flame_oh_hot_man/src/Flame/Renderer/Internal/Dom.js#L634

**Comment by kim366** [comment]
I see, I didn't realize it was this simple

**Comment by kim366** [comment]
There seems to be an issue with DOM-diffing or maybe I have something set up incorrectly. I implemented the example from the project page (replacing `HTML` with `Html`) on the `flame_oh_hot_man` branch and the counter works correctly. The issue is, that whenever I click one of the buttons, the entire subtree of the `<main>` gets updated, including the buttons, which do not change. Is this known?

**Comment by easafe** [comment]
> the entire subtree of the `<main>` gets updated

Not quite sure what you mean by this. The diff'ing should visit all nodes, but only reproduce the changes to the `view`

There is a difference in this branch, however, in that the mount point is no longer replaced. For example if the app is mounted on a `main` tag (like the READ example) then the `main` tag from the view will up the appended to it, resulting in 2 `main`s. I guess the examples should be mounted on the `body` now, but I am not sure if that's completely safe.

**Comment by kim366** [comment]
Yes, that's true, the root is not updated. But its entire subtree is. There are DOM changes to the buttons. In Chrome Devtools you can see it flash when it changes. But maybe the Counter example just has to be updated for this branch

**Comment by easafe** [comment]
If there are DOM changes without `view` changes, then it is likely a bug. I will take a closer look later.

**Comment by kim366** [comment]
I can confirm that the issue was fixed. Thanks for you work!

**Comment by easafe** [comment]
Thanks for spotting it, @kim366 . It turned out to be a silly oversight: text nodes were being updated regardless of changes 😅

**Comment by easafe** [comment]
After adding support for non bubbling events (such as scroll or input), I think synthetic events are not worth it. It complicates the code for not much gain, and the user can always do it manually if needed anyway.

**Comment by chekoopa** [comment]
Bumping the issue to notify that I'm now on my way to bet-tle-test this version. Is it stable enough?

**Comment by easafe** [comment]
I am using it for https://melan.chat and works better (for me) than v0.7.0 (lots of fragments and condtional views). There is an issue with updating SVG nodes in very specific conditons, but I have yet to narrow it down exactly. There is also no functor instance yet.

**Comment by chekoopa** [comment]
Lack of Functor instance is bad, as it's used to encapsulate modules inside each other (using global message type instead seems to be a very bad idea).

And our main Purs-based UI is based on **very** heavy SVG usage.

We'll decide on bet-tle-test this week, then. :thinking:

**Comment by easafe** [comment]
> Lack of Functor instance is bad, as it's used to encapsulate modules inside each other (using global message type instead seems to be a very bad idea)

I have a message type per entry, I find it better than juggling types. But for the functors, it is just a matter of drawing up an efficient way to do it. The update function for a mount is currently cached, but using functors kind of assume they are saved on the node (specially if you want all functor laws to hold). Suggestions welcome.

> And our main Purs-based UI is based on very heavy SVG usage.

You might run into it then. Something like old properties/attributes not being updated whereas new ones work fine. Bizarrely, it only happens for me in a single piece of code.

**Comment by easafe** [comment]
@chekoopa I have added back the function instance. SVGs should also be okay. Could you test it now?

File: src/Flame/Types.purs [hunk 1]
```diff
@@ -38,3 +39,9 @@ foreign import data NodeData :: Type -> Type
 --Html can actually be typed, but since it is only used in FFI code, I don't think it'd be very useful
 -- | The type of virtual nodes
 foreign import data Html :: Type -> Type
+
+--we support events that are not fired on Nothing message
+foreign import messageMapper :: forall message mapped. (Maybe message -> Maybe mapped) -> Html message -> Html mapped
```
Comment (chekoopa) [inline src/Flame/Types.purs:44]: Do we need Maybe to Maybe mapping? Like, there's a possibility to fire an event from Nothing with such a mapper, and its behaviour seems undefined to me. `(msg -> Maybe msg')` is proper here, however, does it comply to functor laws and nature... 🤔

**Review by chekoopa — COMMENTED** [review]
Functor implementation is neat, however something itches me about the function type. I think it should have no Maybe's at all (Nothing is Nothing).

File: src/Flame/Types.purs (hunk 1 above)
Comment (easafe) [reply src/Flame/Types.purs:44]: Hmmm I am not sure that would work. `NodeData` is always `message`, because the `Maybe` value is handled by unsafe JavaScript. See `runHandlers` https://github.com/easafe/purescript-flame/blob/97808aa3cca03cdc1da916e8994a4e07b09bfc70/src/Flame/Renderer/Internal/Dom.js#L362

Maybe you can send a PR to this branch?

Comment (chekoopa) [reply src/Flame/Types.purs:44]: Or we can just double map the value. :laughing:

Comment (easafe) [reply src/Flame/Types.purs:44]: However do you mean? 🤔

Comment (chekoopa) [reply src/Flame/Types.purs:44]: I mean,
```purs
map f html = messageMapper (map (map f)) html
```
It seems cheap, but should work well.

Actually, it really should be `msg -> Maybe mapped`, but it's not a functor behaviour, so it could be implemented with `messageMapper` separately.

Comment (chekoopa) [reply src/Flame/Types.purs:44]: Oh wait... I see, we don't need to map it like this. All right, I was talking about Bind behaviour (we can include it, perhaps), but Functor implementation is all right. 

I'll check it on our SVG UI code ASAP.

Comment (easafe) [reply src/Flame/Types.purs:44]: Did you get the chance to @chekoopa ?

**Comment by easafe** [comment]
I think this one is ready to merge.

**Comment by chekoopa** [comment]
Good job!

Sorry I had no time to properly test it, had to stay and work on Hedwig-based code (as being surely stable). The merge is a good solution, whereas possible SVG issues could be resolved later.

**Comment by easafe** [comment]
Thanks, and no worries, @chekoopa . I think further breaking changes are unlikely, so the last remaining thing to do is getting it on package sets, after is it stable enough.
